runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java

```diff
@@ -235,10 +241,32 @@ private void executeMonitorHeld(Runnable work, long workBytes) {
   }
 
   private void decrementCounters(long workBytes) {
-    monitor.enter();
-    --elementsOutstanding;
-    bytesOutstanding -= workBytes;
-    monitor.leave();
+    // All threads queue decrements and one thread grabs the monitor and updates
+    // counters. We do this to reduce contention on monitor which is locked by
+    // GetWork thread
+    decrementQueue.add(workBytes);
+    synchronized (decrementQueue) {
```

Comment:
might be worth just having a object dedicated to be the lock.  There are warnings usually using synchronized on concurrent/atomic objects about misuse (which I think would be wrong in this case but are still annoying).

Reply:
agree, done.